### ISSUE TITLE
Companion fails loading mastermemories for discrete screens

### ISSUE THREAD

**Issue by bwaack** [bitfocus/companion-module-analogway-livecore#9]
Companion fails loading Mastermemories, if Mastermemory does not not contain presets for every screen at once.

For Example: build mastermemory #1 for screen1 and screen3. Build an other mastermemory (#2) for just screen2.
Recall Mastermemory  #1 in AW-Browser. Companion is now only able to recall Mastermemories which contain presets for screen1 and screen3. Recalling mastermemory #2 in companion will fail - in AW browser it works fine.

After recalling mastermemory #2 in AWbrowser companion will be able to load mastermemory #2 again - but now companion will not load mastermemory #1.
 
Reboot of Ascender solves the issue - but after saving or recalling a mastermemory in AWbrowser, companion is again limited to load only those mastermemories which where configured for the screens recalled in AWbrowser before.

found with:
Ascender 48 4k build 04.02.77
AWbrowser 1.0.19
Companion 2.4.0

**Comment by dnmeid** [comment]
I don't have a device at hand to verify this in the next few days, but honestly I doubt that there is a problem with loading master memories at Companion side. Companion just sends a command to load a memory with a number. Companion really doesn't know what screens are used in that memory or anything else about that memory besides the number. The logic of this module didn't change for over four years and the firmware of Livecore for at least two years. That is the first time I hear of such an issue.
Is it really the recall of the memory which fails or is it the transition of the screens? I often have seen people who don't understand the concept of selecting screens and than they wonder why screens are not transitioning.

**Comment by bwaack** [comment]
Yes, hard to believe that nobody other stumbled about this.Checked it again on differeent machines and computers.Same behavior. No, its not the take. Tried different takes (singe and global) the first time. Finally I deleted all takes. No recall of the described memories.Second time I didnt use any take.There must be a difference between AWbrowser and Companion requesting Mastermemories. I fumbeld around  with screen selection in companion and AWbrowser (screen, Tbarselection, preview or program selected in AWbrowser) No solution found.Strange that it works after reboot of Ascender until you press mastermemories in AWbrowser.GreetsB.

 
  
  
 
 Am 28.11.22, 01:06 schrieb dnmeid ***@***.***>:

   
  I don't have a device at hand to verify this in the next few days, but honestly I doubt that there is a problem with loading master memories at Companion side. Companion just sends a command to load a memory with a number. Companion really doesn't know what screens are used in that memory or anything else about that memory besides the number. The logic of this module didn't change for over four years and the firmware of Livecore for at least two years. That is the first time I hear of such an issue. Is it really the recall of the memory which fails or is it the transition of the screens? I often have seen people who don't understand the concept of selecting screens and than they wonder why screens are not transitioning.

**Comment by dnmeid** [comment]
Technically there is a difference in the transmission of the data with AWbrowser and the TCP/IP protocol used by Companion, but the command set is the same. As I said if this issue is reproducable, I don't know what we could do about it. Companion just sends a command like: load mastermemory 2 to preview. If this does not work only if there are certain screens contained in the memory or not contained in the memory, or what has previously done in AWbrowser it has to be an issue of LiveCore. Companion always sends the same command and doesn't know about the screens affected or what you do in AWbrowser or if this is the first recall after boot.

**Comment by dnmeid** [comment]
Do you had by any chance during your tests always had the programm locked and tried only to load a memory to preview? If yes can you try the same with program unlocked and see if maybe the load wants to go to the wrong preset for the screen missing in the last recall.

**Comment by dnmeid** [comment]
I just had a look at the documentation of the protocol and there is an optional filter where one can exclude some screens from loading a master memory. In the documentation it says, that if this filter is omitted (and Companion omits it) no filtering will take place. I can imagine that AWbrowser sets this filter for the screens it wants to load and then this filter also applies to the load of Companion which it should not.
Can you try the following:
- load MM1 from AWbrowse
- execute custom Command 2,2,1PSsse (this should explicitly disable the filter for MM2 on screen 2)
- load MM2 from Companion

**Comment by bwaack** [comment]
ah, that would exactly explain the behavior.AW browser does not selcect screens for loading, but filters loading for screens not selected in mastermemory. But it does not reset the filter after preset is loaded.So next loading which is not using the filter will only affect screens which where loaded before from AWbrowser.So you have to set the filter before every recall from companion to load every mastermemory for every screen for sure from companion.I will try that, next time… if there is time for that.Which number stands for screen, which for memory?This filter is always set for a special MM? No „dont load screen2“?Is it necessary to set delays after those commands? How much?By the way, I just checked, the first time i saw the issue with Ascender 32. So probably all Ascenders are affected. May be there is a way to sniff the code AWbrowser sends.Thanks for your workB.Am 29.11.22 um 00:30 schrieb dnmeid
                
            
            
                
                    Von: "dnmeid" ***@***.***>Datum: 29. November 2022An: "bitfocus/companion-module-analogway-livecore" ***@***.***>Cc: "Author" ***@***.***>,"bwaack" ***@***.***>Betreff: Re: [bitfocus/companion-module-analogway-livecore] Companion fails loading mastermemories for discrete screens (Issue #9)
                
                
I just had a look at the documentation of the protocol and there is an optional filter where one can exclude some screens from loading a master memory. In the documentation it says, that if this filter is omitted (and Companion omits it) no filtering will take place. I can imagine that AWbrowser sets this filter for the screens it wants to load and then this filter also applies to the load of Companion which it should not.
Can you try the following:

load MM1 from AWbrowse
execute custom Command 2,2,1PSsse (this should explicitly disable the filter for MM2 on screen 2)
load MM2 from Companion

**Comment by dnmeid** [comment]
@Enga18 No, bwaak stopped responding and noone else could reproduce this problem so far or was able to collaborate to find the reason.

**Comment by Enga18** [comment]
@dnmeid thank you so much..! Hope someone will find a solution. @bwaack I belive in you!

**Comment by bwaack** [comment]
Yes, dnmeid, it is the filter!

If AW-Browser recalls a MM with only some screens, not all, it leaves the filter for not recalled screens - so companion will fail for future recalls on those screens, which where not recalled by AW-Browser.

The custom Command "PSsse" changes nothing, because it is a read-only one.
But there is an other custom command: PSose!

So: If companion fails recalling MMs after you pressed a MM in AW-Browser you can:

Press an MM in AW-Browser which addresses all screens, companion will work again after that.

or:

send custom commands before recalling MMs in companion.

0,1PSose will activate screen1 again
1,1PSose will activate screen2 again
2,1Psose will activate screen3 again 
and so forth.

For security reasons you may send those PSose-commands for all your screens before every recall of MM in Companion.

greetz
b.
